## #1, #2 & #3. Creating path shortcut and importing libraries & dataframe

In [2]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
pathname=r'/Users/alejandrocisneros/Desktop/Data Analysis Coursework/Data Immersion/Achievement 4/02-2024 Instacart Basket Analysis'

In [4]:
df=pd.read_csv(os.path.join(pathname, '2. Data', 'Original Data', 'customers.csv'))

## #4 & #5. Wrangle the data so that it follows consistent logic and ensure data quality and consistency

In [5]:
#Preview of the data
df

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

## #Noted... 206k customers in the database.

In [6]:
df.shape

(206209, 10)

In [7]:
df.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## #First_Name column has a lower value count. Lets verify

In [9]:
df['First Name'].count()

194950

In [10]:
df['First Name'].value_counts(dropna=False)

First Name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

## #Given that the column missing values is a string, we cannot impute values. This will always be reviewed with a supervisor or stakeholder in a real situation, but we will delete all 11,259 rows where the first name is missing.

In [11]:
df=df[df['First Name'].notnull()]

In [12]:
#Testing results
df['First Name'].value_counts(dropna=False)

First Name
Marilyn    2213
Barbara    2154
Todd       2113
Jeremy     2104
Cynthia    1951
           ... 
Merry       197
Eugene      197
Garry       191
Ned         186
David       186
Name: count, Length: 207, dtype: int64

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 194950 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       194950 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        194950 non-null  object
 3   Gender        194950 non-null  object
 4   STATE         194950 non-null  object
 5   Age           194950 non-null  int64 
 6   date_joined   194950 non-null  object
 7   n_dependants  194950 non-null  int64 
 8   fam_status    194950 non-null  object
 9   income        194950 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 16.4+ MB


In [14]:
#Now, we are changing the data type for the user_id column to a string, given that its values are categorical, not 
#measurable
df['user_id']=df['user_id'].astype('str')

/var/folders/rg/4c63ck7s21b88c7996mzk3cw0000gn/T/ipykernel_72323/190519282.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['user_id']=df['user_id'].astype('str')


In [15]:
df.describe()

Age   n_dependants         income
count  194950.000000  194950.000000  194950.000000
mean       49.511167       1.499600   94664.811495
std        18.476519       1.117959   42477.262653
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59924.000000
50%        49.000000       1.000000   93572.500000
75%        66.000000       2.000000  124261.000000
max        81.000000       3.000000  593901.000000

In [16]:
#Data has been wrangled. No columns need intuitive renaming, nor do they need data type changing. No tables need
#transposing so we are done with wrangling

In [17]:
#First, we are going to write similar code to previous exercises to identify if there are any columns with mixed
#data types with a for loop and if statement, finding it thru the assignment of Boolean values to determine whether
#or not the condition is met
for col in df.columns.tolist():
  weird = (df[[col]].map(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

In [18]:
#No columns with mixed data types in this dataframe

In [19]:
#Now, we will check for duplicated rows in the data frame by using the duplicated() func
df[df.duplicated()]

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []

In [20]:
#No duplicates in the data frame either so we are all set in the data quality and consistency checks

# #6.Combine customer data with revised Instacart data

In [21]:
##We could or should have done this in the first section, but we will now import the Instacart dataframe in order
##to merge it with the customer set
ords_prods_revised=pd.read_pickle(os.path.join(pathname, '2. Data', 'Prepared Data', 'ords_prods_merge_final.pkl'))

In [22]:
#Testing
ords_prods_revised.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered      price_segment    busiest_days   busiest_hours  \
0  ...          0  Mid_range product  Regularly busy     Most orders   
1  ...          1  Mid_range product  Regularly busy  Average_orders   
2  ...          0  Mid_range product     Busiest day  Average_orders   
3  ...          0  Mid_range product      Least busy  Average_orders   
4  ...          1  Mid_range product      Least busy  Average_orders   

  max_order      loyalty_flag  avg_prod_purchase_price spender_tier  \
0        32  Regular Customer                 6.935811  Low Spender   
1        32  Regular Customer                 6.935811  Low Spender   
2         5      New Customer                 7.930208  Low Spender   
3         3      New Customer                 4.972414  Low Spender   
4         3      New Customer                 4.972414  Low Spender   

   median_reorder_time     frequency_flag  
0             8.000000  Frequent Customer  
1             8.000000  Frequent Customer  
2             8.000000  Frequent Customer  
3            11.114836   Regular Customer  
4            11.114836   Regular Customer  

[5 rows x 22 columns]

In [24]:
#In order to have the same data types for the key column, we will need to change the data type for the 'user_id'
#column in the ords_prods_final dataframe to match the object data type we applied to the same column in the 
#customer 'df' dataframe
ords_prods_revised['user_id']=ords_prods_revised['user_id'].astype('str')

In [25]:
#Testing if changing column data type worked
ords_prods_revised.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 22 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   product_id               int64  
 1   product_name             object 
 2   aisle_id                 int64  
 3   department_id            int64  
 4   prices                   float64
 5   order_id                 int64  
 6   user_id                  object 
 7   order_number             int64  
 8   orders_day_of_week       int64  
 9   order_hour_of_day        int64  
 10  days_since_last_order    float64
 11  add_to_cart_order        int64  
 12  reordered                int64  
 13  price_segment            object 
 14  busiest_days             object 
 15  busiest_hours            object 
 16  max_order                int64  
 17  loyalty_flag             object 
 18  avg_prod_purchase_price  float64
 19  spender_tier             object 
 20  median_reorder_time      float64
 21  freque

### #Success!

In [26]:
##We will combine this customer dataframe with the revised ords_prods_merge dataframe containing the majority of 
##information on Instacart data
instacart_final=df.merge(ords_prods_revised, on='user_id', indicator=True)

In [27]:
instacart_final

user_id First Name    Surnam  Gender     STATE  Age date_joined  \
0          26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1          26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2          26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3          26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4          26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
...          ...        ...       ...     ...       ...  ...         ...   
30629736   80148    Cynthia     Noble  Female  New York   55    4/1/2020   
30629737   80148    Cynthia     Noble  Female  New York   55    4/1/2020   
30629738   80148    Cynthia     Noble  Female  New York   55    4/1/2020   
30629739   80148    Cynthia     Noble  Female  New York   55    4/1/2020   
30629740   80148    Cynthia     Noble  Female  New York   55    4/1/2020   

          n_dependants fam_status  income  ...      price_segment  \
0                    3    married  165665  ...  Mid_range product   
1                    3    married  165665  ...  Mid_range product   
2                    3    married  165665  ...  Mid_range product   
3                    3    married  165665  ...  Low_range product   
4                    3    married  165665  ...  Low_range product   
...                ...        ...     ...  ...                ...   
30629736             1    married   57095  ...  Low_range product   
30629737             1    married   57095  ...  Low_range product   
30629738             1    married   57095  ...  Low_range product   
30629739             1    married   57095  ...  Low_range product   
30629740             1    married   57095  ...  Mid_range product   

            busiest_days   busiest_hours  max_order  loyalty_flag  \
0            Busiest day  Average_orders          8  New Customer   
1         Regularly busy     Most orders          8  New Customer   
2            Busiest day     Most orders          8  New Customer   
3         Regularly busy  Average_orders          8  New Customer   
4             Least busy     Most orders          8  New Customer   
...                  ...             ...        ...           ...   
30629736  Regularly busy     Most orders          4  New Customer   
30629737  Regularly busy     Most orders          4  New Customer   
30629738  Regularly busy     Most orders          4  New Customer   
30629739  Regularly busy     Most orders          4  New Customer   
30629740  Regularly busy     Most orders          4  New Customer   

          avg_prod_purchase_price  spender_tier  median_reorder_time  \
0                        7.988889   Low Spender                 19.0   
1                        7.988889   Low Spender                 19.0   
2                        7.988889   Low Spender                 19.0   
3                        7.988889   Low Spender                 19.0   
4                        7.988889   Low Spender                 19.0   
...                           ...           ...                  ...   
30629736                 3.886667   Low Spender                 12.0   
30629737                 3.886667   Low Spender                 12.0   
30629738                 3.886667   Low Spender                 12.0   
30629739                 3.886667   Low Spender                 12.0   
30629740                 3.886667   Low Spender                 12.0   

            frequency_flag  _merge  
0         Regular Customer    both  
1         Regular Customer    both  
2         Regular Customer    both  
3         Regular Customer    both  
4         Regular Customer    both  
...                    ...     ...  
30629736  Regular Customer    both  
30629737  Regular Customer    both  
30629738  Regular Customer    both  
30629739  Regular Customer    both  
30629740  Regular Customer    both  

[30629741 rows x 32 columns]

In [28]:
#Checking the info() function to verify integrity before exporting
instacart_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30629741 entries, 0 to 30629740
Data columns (total 32 columns):
 #   Column                   Dtype   
---  ------                   -----   
 0   user_id                  object  
 1   First Name               object  
 2   Surnam                   object  
 3   Gender                   object  
 4   STATE                    object  
 5   Age                      int64   
 6   date_joined              object  
 7   n_dependants             int64   
 8   fam_status               object  
 9   income                   int64   
 10  product_id               int64   
 11  product_name             object  
 12  aisle_id                 int64   
 13  department_id            int64   
 14  prices                   float64 
 15  order_id                 int64   
 16  order_number             int64   
 17  orders_day_of_week       int64   
 18  order_hour_of_day        int64   
 19  days_since_last_order    float64 
 20  add_to_cart_order     

# #7, #8 & #9. Ensure notebook cleanliness and export into pickle file

In [29]:
#Exporting
instacart_final.to_pickle(os.path.join(pathname, '2. Data', 'Prepared Data', 'final_instacart.pkl'))